enhancements/local-storage/automatic-detection.md

```diff
@@ -0,0 +1,320 @@
+
+## Design Details for `Auto provisioning of local devices`
+API scheme for `LocalVolumeGroups`:
+
+```go
+type LocalVolumeGroupList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+	Items           []LocalVolumeGroup `json:"items"`
+}
+
+type LocalVolumeGroup struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Spec              LocalVolumeGroupSpec   `json:"spec"`
+	Status            LocalVolumeGroupStatus `json:"status,omitempty"`
+}
+
+type LocalVolumeGroupSpec struct {
+	// Nodes on which the automatic detection policies must run.
+	// +optional
+	NodeSelector *corev1.NodeSelector `json:"nodeSelector,omitempty"`
+	// StorageClassName to use for set of matched devices
+	StorageClassName string `json:"storageClassName"`
+	// MinDeviceCount is the minumum number of devices that needs to be detected per node.
+	// If the match devices are less than the minCount specified then no PVs will be created.
+	MinDeviceCount int `json:"minDeviceCount"`
+	// +optional
+	// Maximum number of Devices that needs to be detected per node.
+	MaxDeviceCount int `json:"maxDeviceCount"`
+	// VolumeMode determines whether the PV created is Block or Filesystem. By default it will
+	// be block
+	// + optional
+	VolumeMode PersistentVolumeMode `json:"volumeMode,omitempty"`
+	// FSType type to create when volumeMode is Filesystem
+	// +optional
+	FSType string `json:"fsType,omitempty"`
+	// If specified, a list of tolerations to pass to the discovery daemons.
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
+	// DeviceInclusionSpec is the filtration rule for including a device in the device discovery
+	// +optional
+	DeviceInclusionSpec *DeviceInclusionSpec `json:"deviceInclusionSpec"`
+}
+
+type LocalVolumeGroupStatus struct {
+	Status string `json:"status"`
+}
+
+// DeviceMechanicalProperty holds the device's mechanical spec. It can be rotational or nonRotational
+type DeviceMechanicalProperty string
+
+const (
+	// The mechanical properties of the devices
+	// Rotational refers to magnetic disks
+	Rotational DeviceMechanicalProperty = "Rotational"
+	// NonRotational refers to ssds
+	NonRotational DeviceMechanicalProperty = "NonRotational"
+)
+
+type DeviceDiscoveryPolicyType string
+
+const (
+	// The DeviceDiscoveryPolicies that will be supported by the LSO.
+	// These Discovery policies will based on lsblk's type output
+	// Disk represents a device-type of disk
+	Disk DeviceDiscoveryPolicyType = "disk"
+	// Part represents a device-type of partion
+	Part DeviceDiscoveryPolicyType = "part"
+)
+
+type DeviceInclusionSpec struct {
+	// DeviceTypes that should be used for automatic detection. This would be one of the types supported
+	// by the local-storage operator. Currently, the supported types are: disk, part
+	// If the list is empty the default value will be `[disk]`.
```

Comment:
```I would propose, if nil, use default (different in each field). If empty array, don't match anything - user explicitly set the field to empty array (which is not nil!), so obey it.```
okay instead of []DeviceType, I created one struct called `DeviceTypeList` and the `DeviceInclusionSpec` holds the pointer to that. So now if the `DeviceTypeList` is nil we will just be detecting the `Disk` type of devices